### PR TITLE
Drop support for python < 3.5

### PR DESCRIPTION
**Pull request by a-detiste**

### CODE REVIEW

**Comment by AdamWill** [comment]
Thanks! As you can see, though, needs some fixes. Yay for tests :)

I was planning to do a kinda more extensive cut at this - use more modern features like f-strings - but I never got around to it, so I'm fine with a more minimal version, if you can fix the failures. Thanks again.

**Comment by a-detiste** [comment]
annotations + mypy is a more usefull modern feature than f-string... I think

**Comment by AdamWill** [comment]
sure, that would be useful too!

**Comment by a-detiste** [comment]
>  that would be useful too!

That would be usefull _right_ _now_ ,
it would have told me I'm trying to stuff unexpected content in `self._iter`

If the test fails again I'll do that.

If you want it I can do it anyway in an other PR.

**Comment by AdamWill** [comment]
I am all for type annotations, seems like it'd be quite a big job for this library though?

This LGTM now, let's give it a bit of time for other maintainers to take a look too.

**Comment by a-detiste** [comment]
Yes, one thing at a time :-) 

The framework evolves fast,
so you might after a while decide that 3.5 is too old as a baseline.

It can be iterative, with a pretty good cost/gain ratio.

Here for example I dig up what was a 'site:',
bot not yet 'return_values:'.


```
--- a/mwclient/listing.py
+++ b/mwclient/listing.py
@@ -2,6 +2,9 @@ from mwclient.util import parse_timestamp
 import mwclient.page
 import mwclient.image
 
+from typing import Optional, TYPE_CHECKING
+if TYPE_CHECKING:
+    from mwclient.client import Site
 
 class List:
     """Base class for lazy iteration over api response content
@@ -11,9 +14,17 @@ class List:
     continuing content.
     """
 
-    def __init__(self, site, list_name, prefix,
-                 limit=None, return_values=None, max_items=None,
-                 *args, **kwargs):
+    def __init__(
+        self,
+        site: 'Site',
+        list_name: str,
+        prefix: str,
+        limit: Optional[int] = None,
+        return_values=None,
+        max_items: Optional[int] = None,
+        *args,
+        **kwargs
+    ) -> None:
         # NOTE: Fix limit
         self.site = site
         self.list_name = list_name
```

**Comment by AdamWill** [comment]
Personally I tend to use 3.6 as the baseline as I work for Red Hat and it's in the system Python in some RHEL or other :P But this project is kinda collaborative - the original maintainers have all gone inactive, now a few of us who are active users of the library are trying to take care of it. So we'll have to figure it out together.

**Comment by marcfrederick** [comment]
This looks great, thank you @a-detiste! Typing support will be great for this library and this will help us get there.

As for the minimum version, I think I'd also drop 3.5 while we're at it to get f-strings and variable annotations.

**Comment by NguoiDungKhongDinhDanh** [comment]
_tox.ini_ also needs to be modified so as to remove `py27` (as well as `34` and `35`, if we were to do this thoroughly) from `envlist`:

https://github.com/a-detiste/mwclient/blob/master/tox.ini#L2

**Comment by a-detiste** [comment]
so this is done ?

File: examples/upload.py [hunk 1]
```diff
@@ -1,4 +1,3 @@
-from __future__ import print_function
 from past.builtins import xrange
 from StringIO import StringIO
```
Comment (marcfrederick) [inline examples/upload.py:2]: The `StringIO` module [is not available in Python 3](https://docs.python.org/3.5/whatsnew/3.0.html?highlight=stringio):

> The StringIO and cStringIO modules are gone. Instead, import the [io](https://docs.python.org/3.5/library/io.html#module-io) module and use [io.StringIO](https://docs.python.org/3.5/library/io.html#io.StringIO) or [io.BytesIO](https://docs.python.org/3.5/library/io.html#io.BytesIO) for text and data respectively.

```suggestion
from io import StringIO
```

File: examples/upload.py [hunk 2]
```diff
@@ -1,4 +1,3 @@
-from __future__ import print_function
 from past.builtins import xrange
```
Comment (marcfrederick) [inline examples/upload.py:1]: This relies on the `future` module (which is unavailable), we should use the builtin `range` instead.

**Review by marcfrederick — CHANGES_REQUESTED** [review]
Hey @a-detiste thanks again for you're contribution. I found two more small issues in the `upload.py` example, apart from that this looks good to me and I would be in favor of getting this merged.

**Comment by a-detiste** [comment]
I pushed more changes, I guess it's better to squash everything in the merge.

**Review by marcfrederick — APPROVED** [review]

**Comment by marcfrederick** [comment]
Thanks, from my side, this looks good now. The baseline is still 3.5 but, I think we can defer the decision on whether to drop that in favor of 3.6 to another discussion and proceed with getting this one merged.

**Comment by AdamWill** [comment]
Can we update the commit message on the first commit from 3.3 to 3.5, in case that gets used however we merge this?

**Comment by AdamWill** [comment]
other than that I guess this looks good.

**Comment by a-detiste** [comment]
done

**Review by AdamWill — APPROVED** [review]

**Comment by a-detiste** [comment]
Ping

**Comment by AdamWill** [comment]
Sorry, have not had a lot of time for mwclient stuff lately with $DAYJOB :( merging, thanks!

**Comment by AdamWill** [comment]
ugh, actually - @marcfrederick , I don't remember what was the merge strategy we agreed on. For my own projects I use rebase but I remember you didn't like that. Are we using squash and merge?

**Comment by marcfrederick** [comment]
No, I also prefer using rebase and tend to use that for my personal projects. I've only used merge here because I saw you using that. Let's just go with rebase merging (and possibly enforce that in the settings?).

**Comment by AdamWill** [comment]
ugh, jeez, well certainly *someone* wanted to use a different merge strat. maybe @waldyrious ?

**Comment by AdamWill** [comment]
Well, it's been two weeks and the only maintainers who replied like rebasing, so let's rebase. :D Thanks again for the work!

**Comment by waldyrious** [comment]
> ugh, jeez, well certainly _someone_ wanted to use a different merge strat. maybe @waldyrious ?

Oh yeah, [that was me](https://github.com/mwclient/mwclient/pull/281#issuecomment-1426376607) — sorry it took me so long to reply here. So, for future reference, the strategy in question consists in rebasing a branch (if it's not based on the latest commit of the target branch) and then merging via a merge commit (instead of via squashing or a fast-forward merge).

This method —explained [here](https://github.com/isaacs/github/issues/1017) in more detail, including a helpful diagram— is sometimes named "semi-linear", as it generates a nearly linear git history, while still making use of merge commits that allow multiple commits from a branch to be grouped when navigating the history. (Of course, it's only useful to preserve the separate commits if they are indeed atomic, deliberately separated and well-described.

In any case, that was just my personal preference; it never was encoded in the project's contribution documentation, so unless you like the idea, feel free to continue using the rebase + fast-forward merge method.

**Pull request merged**